src/matcha_ml/services/global_config_service.py

```diff
@@ -0,0 +1,139 @@
+"""Global config service for creating and modifying a users global config files."""
+import os
+import uuid
+from typing import Any, Dict, Optional
+
+import yaml
+
+from matcha_ml.errors import MatchaPermissionError
+
+
+class GlobalConfigurationService:
+    """A Global Config Service for interacting and updating a users global config file."""
+
+    _instance: Optional["GlobalConfigurationService"] = None
+    _user_id: Optional[str] = None
+    _analytics_opt_out: bool = False
+
+    def __new__(cls) -> "GlobalConfigurationService":
+        """Singleton class definition.
+
+        Returns:
+            GlobalConfigurationService: Already existing initialised object, otherwise a new singleton object
+        """
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+            # Check if config.yaml file exists and read in variables to the class
+            if os.path.exists(str(cls._instance.default_config_file_path)):
+                cls._instance._read_global_config(
+                    str(cls._instance.default_config_file_path)
+                )
+            else:
+                cls._instance._create_global_config(
+                    config_file_path=str(cls._instance.default_config_file_path)
+                )
+
+        return cls._instance
+
+    def _read_global_config(self, config_file_path: str) -> None:
+        """Reads the global config yaml file.
+
+        Args:
+            config_file_path (str): Path to users global config file
+        """
+        with open(config_file_path) as file:
+            yaml_data = yaml.safe_load(file)
+            self._user_id = yaml_data.get("user_id")
+            self._analytics_opt_out = yaml_data.get("analytics_opt_out") == "True"
```

Comment:
Do we need `== True` here?

Reply:
Why are the following lines within the context manager of `open`?

```python
self._user_id = yaml_data.get("user_id")
self._analytics_opt_out = yaml_data.get("analytics_opt_out") == "True"
```

Reply:
`== "True"` is required so that is it parsed into a bool. Otherwise, it will be a string.

Reply:
I think we can store booleans in yaml. They can be either `true` or `false`. Is there a reason why we store them as string instead of boolean?